### ISSUE TITLE
GPU (backend cuda) build problem

### ISSUE THREAD

**Issue by iamyoungjo** [clab/dynet#300]
I am having problem on build with BACKEND=cuda. My system is OS X (10.11.6 El Capitan)
cmake works fine, once I do "make -j 4", it returns following error:
>Undefined symbols for architecture x86_64:
> ...
>ld: symbol(s) not found for architecture x86_64
>clang: error: linker command failed with exit code 1 (use -v to see invocation)

As I use GPU in tensorflow without any issues, I doubt that my CUDA setting has not done or something.

I searched similar issues here but I am the only person who's having this issue as it seems.
There's no issue on compilation if I don't put cuda backend.

If I missed a significant step here or if anyone who's familiar with this error, please help.
I wasted more than 6 hours because of this.

[make.log.zip](https://github.com/clab/dynet/files/771468/make.log.zip)

**Comment by neubig** [comment]
I was able to replicate this, so I'll try to figure out the problem.

**Comment by neubig** [comment]
This is potentially fixed by https://github.com/clab/dynet/pull/301

Could you try this out and see if this fixes the problems?

**Comment by iamyoungjo** [comment]
I just tried, and it shows a different error:
> ld: library not found for -lcublas
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> make[2]: *** [dynet/libdynetcuda.dylib] Error 1
> make[1]: *** [dynet/CMakeFiles/dynetcuda.dir/all] Error 2
> make: *** [all] Error 2

[make.log.zip](https://github.com/clab/dynet/files/771669/make.log.zip)

**Comment by iamyoungjo** [comment]
Any possibility that I am missing any library? Here'e my sample run for cuda setting validity check.
[cuda_test.log.zip](https://github.com/clab/dynet/files/771711/cuda_test.log.zip)

**Comment by iamyoungjo** [comment]
@neubig any ideas? I am still stuck in there..

**Comment by xunzhang** [comment]
@iamyoungjo According to your log, the reason why it is still failing is that the default linking args `-lcublas` could only search in library files from`/usr/lib` `/usr/local/lib`, while the default path for cuda is `/usr/local/cuda/lib`.

A work around should be expending search path using `DYLD_LIBRARY_PATH` under osx, like this:
`export CUDA_HOME="/usr/local/cuda"`
 `export DYLD_LIBRARY_PATH="/usr/local/cuda/lib"`
To point our, for OSX >10.11, you need to turning off Rootless System Integrity Protection to honor `DYLD_LIBRARY_PATH` variable. Follow this [link](http://osxdaily.com/2015/10/05/disable-rootless-system-integrity-protection-mac-os-x/)

@neubig I am now creating a pull request to fix this issue. To give an error message or use `-Wl,-rpath`.

**Comment by xunzhang** [comment]
The default cmake finding results for `CUDA_LIBRARIES` is only cuda rt libraries. https://github.com/clab/dynet/blob/master/CMakeLists.txt#L146

**Comment by neubig** [comment]
This should be fixed in master by https://github.com/clab/dynet/pull/303

**Comment by xunzhang** [comment]
@neubig I haven't tested in my linux environment yet, I will do that asap. It should work. I think we should add CI with cuda configuration but I need to test the time to make sure it is not too long. To install cuda, I also need a public place download cuda and cudnn libraries. Let's discuss in another thread later.

**Comment by neubig** [comment]
@xunzhang It worked on my linux. And +1 to adding cuda compiling to CI.

**Comment by xunzhang** [comment]
Good to know that. I will do that tmr. @iamyoungjo have a try and please close the issue if latest code solved your problem, thanks.

**Comment by iamyoungjo** [comment]
First of all, thank you so much for you guys' effort. I just tried and was possible to go through up to 99% but it omitted one error at the last moment.

> ld: library not found for -llibcublas.dylib
> clang-3.9: error: linker command failed with exit code 1 (use -v to see invocation)
>error: command '/usr/local/opt/llvm/bin/clang++' failed with exit status 1
> make[2]: *** [python/timestamp] Error 1
> make[1]: *** [python/CMakeFiles/target.dir/all] Error 2

[make2.log.zip](https://github.com/clab/dynet/files/773048/make2.log.zip)

**Comment by xunzhang** [comment]
I could successfully build under osx 10.12. You are using osx 10.6?

**Comment by iamyoungjo** [comment]
I am using 10.11.6 (El Capitan).

**Comment by iamyoungjo** [comment]
@xunzhang I tried with different compiler but still it gives me the same error. Any workaround or fix to deal with this? I spent whole day with this.

**Comment by neubig** [comment]
@iamyoungjo Thanks for the patience! I understand the frustration, but also please understand that it's hard for us to debug when we don't have access to a machine that can replicate the problem.

Anyway, it looks like this problem is now stemming from CUBLAS problems when compiling the python library. The relevant lines that are setting the CUBLAS path are here:
https://github.com/clab/dynet/blob/master/python/setup.py.in#L36
https://github.com/clab/dynet/blob/master/python/CMakeLists.txt#L27

Probably what will need to be done to fix the problem is taking a look at `make2.log`, comparing how cublas is linked when compiling is working (in the main DyNet directory), comparing this with when it's not working (in the Python directory), and attempting to resolve the difference by modifying `setup.py.in` or `CMakeLists.txt`. I won't have time to do this today, but if someone wants to take a shot it'd be appreciated.

**Comment by xunzhang** [comment]
I will take a look at it today, finding an osx 10.11 environment and try to reproduce the bug. @iamyoungjo What's your email address, I want head up to you asking more details in a more interactive way. Thanks.

**Comment by iamyoungjo** [comment]
@xunzhang dearyoungjo@gmail is my email address. Thanks!

**Comment by xunzhang** [comment]
I will summarize the offline discussion with @iamyoungjo through Gmail.

I found the issue came out when specifying `-DPYTHON` arg during configuration under osx. After adding this option, I can replicate the bug in my environment. This is another issue not the same as the original proposal one, the root cause for this is that dynet doesn't support python build under osx.

I have fixed it at https://github.com/clab/dynet/pull/305
